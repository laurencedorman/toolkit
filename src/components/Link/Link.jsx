/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styles from './Link.module.scss';

const Link = ({ onClick, href, children, ...restOfProps }) => {
  const tagName = !onClick && href ? 'a' : 'button';
  const buttonProps = tagName === 'button' ? { type: 'button' } : {};

  return React.createElement(
    tagName,
    {
      className: styles.Link,
      href,
      onClick,
      ...buttonProps,
      ...restOfProps,
    },
    children
  );
};

const requiredPropsCheck = (props, propName, componentName) => {
  if (!props.href && !props.onClick) {
    return new Error(
      `One of 'href' or 'onClick' is required by ${componentName} component.`
    );
  }
};

Link.propTypes = {
  href: requiredPropsCheck,
  onClick: requiredPropsCheck,
};

export default Link;
