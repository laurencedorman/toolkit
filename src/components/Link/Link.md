```js
<Link href="https://www.manomano.fr">ManoMano</Link>
```

```js
const onClick = () => alert('You clicked the link!');

<Link onClick={onClick}>Click me</Link>;
```

```js
<Link disabled={true} onClick={() => null}>
  Click me
</Link>
```
