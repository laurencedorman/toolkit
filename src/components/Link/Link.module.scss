/**
 * Copyright (c) Colibri SAS - ManoMano
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@import '../../styles/import.scss';

.Link {
  appearance: none;
  text-decoration: none;
  border: none;
  padding: 0;
  font-size: em(16);
  color: $balataGreen;
  font-family: $main-font;
  cursor: pointer;

  &:active {
    color: $blueMood;
  }

  &:visited {
    color: $purpleRain;
  }

  &:disabled {
    cursor: not-allowed;
    color: $greyDark;
  }

  &:hover {
    text-decoration: underline;
  }
}
