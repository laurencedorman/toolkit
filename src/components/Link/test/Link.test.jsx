import React from 'react';
import { shallow } from 'enzyme';
import Link from '../Link';

describe('Link', () => {
  it('should display an anchor tag if href in props', () => {
    const wrapper = shallow(<Link href="http://google.fr">Google</Link>);

    expect(wrapper.type()).toEqual('a');
    expect(wrapper).toMatchSnapshot();
  });

  it('should display a button tag if onClick in props', () => {
    const wrapper = shallow(<Link onClick={() => null}>Click me</Link>);

    expect(wrapper.type()).toEqual('button');
    expect(wrapper.find('[type="button"]')).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
  });

  it('should pass through props correctly', () => {
    const wrapper = shallow(
      <Link target="_blank" alt="Some alt text" href="http://google.fr">
        Google
      </Link>
    );

    expect(wrapper.find('[target="_blank"]')).toHaveLength(1);
    expect(wrapper.find('[alt="Some alt text"]')).toHaveLength(1);
  });
});
