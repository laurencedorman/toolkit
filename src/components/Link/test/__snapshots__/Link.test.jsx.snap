// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should display a button tag if onClick in props 1`] = `
<button
  className="Link"
  onClick={[Function]}
  type="button"
>
  Click me
</button>
`;

exports[`Link should display an anchor tag if href in props 1`] = `
<a
  className="Link"
  href="http://google.fr"
>
  Google
</a>
`;
